### PR TITLE
Enable Pipeline Parallelism support for Piecewise CUDA Graph #14515

### PR DESCRIPTION
**Pull request by baonudesifeizhai**
## Motivation
Previously, piecewise CUDA graph was explicitly disabled when pipeline parallelism (PP) was enabled. This PR enables PP support for piecewise CUDA graph, allowing users to benefit from both optimizations simultaneously.

#14515
## Modifications
- Removed the PP size check that disabled piecewise CUDA graph in `ModelRunner.can_run_piecewise_cuda_graph()`
- Added PP proxy tensors buffer initialization in `PiecewiseCudaGraphRunner.__init__()`
- Added PP proxy tensors handling in `PiecewiseCudaGraphRunner.capture_one_batch_size()` for graph capture
- Added PP proxy tensors handling in `PiecewiseCudaGraphRunner.replay()` to properly pass tensors from pre-allocated buffers during replay
- Cached model signature check to avoid expensive `inspect.signature()` calls in the hot path

## Accuracy Tests
https://paste.ubuntu.com/p/KWKwszGRvd/
res:
] image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [x] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

**Review by ispobock — COMMENTED** [review]
Could you add an unit test?

**Comment by ispobock** [comment]
@baonudesifeizhai Could you resolve the conflict?

File: python/sglang/srt/model_executor/model_runner.py [hunk 1]
```diff
@@ -1678,13 +1678,7 @@ def can_run_piecewise_cuda_graph(self):
...
             )
             return False
-        if self.pp_size > 1:
-            # TODO(yuwei): support PP
-            log_info_on_rank0(
-                logger,
-                "Disable piecewise CUDA graph because piecewise_cuda_graph does not support PP",
-            )
-            return False
+        # PP support is now enabled for piecewise CUDA graph
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/model_runner.py]: remove this line

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py [hunk 2]
```diff
@@ -238,9 +239,34 @@ def __init__(self, model_runner: ModelRunner):
                     (self.max_num_tokens, self.model_runner.model_config.hidden_size),
                     dtype=self.model_runner.dtype,
                 )
-                self.mrope_positions = torch.zeros(
-                    (3, self.max_num_tokens), dtype=torch.int64
-                )
+            self.mrope_positions = torch.zeros(
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: Does PP need mrope or not? if not allocate the buffer when is_multimodal is True

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py [hunk 3]
```diff
@@ -503,6 +543,7 @@ def run_once():
     def replay_prepare(
         self,
         forward_batch: ForwardBatch,
+        pp_proxy_tensors: Optional[PPProxyTensors] = None,
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: Since you do not use `pp_proxy_tensors` please remove it

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py [hunk 4]
```diff
@@ -605,7 +646,29 @@ def replay(
     ) -> Union[LogitsProcessorOutput, PPProxyTensors, EmbeddingPoolerOutput]:
         with enable_piecewise_cuda_graph(), disable_ca_comm(self.model_runner.tp_group):
             self.model_runner.attn_backend.init_forward_metadata(forward_batch)
-            static_forward_batch = self.replay_prepare(forward_batch, **kwargs)
+            # Extract pp_proxy_tensors from kwargs if present (avoid in-place modification)
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: With original `replay_prepare` there is no need to split `pp_proxy_tensors` from `kwargs`

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py [hunk 5]
```diff
@@ -238,9 +239,34 @@ def __init__(self, model_runner: ModelRunner):
...
+            )
+
+            # PP proxy tensors buffer (for pipeline parallelism)
+            if self.pp_size > 1:
+                hidden_size = self.model_runner.model_config.hidden_size
+                dtype = self.model_runner.model_config.dtype
+                # Note: piecewise captures with bs=1, but we need buffer for PP proxy tensors
+                # The buffer size is 1 since we capture with batch_size=1
+                self.pp_proxy_tensors_buffer = {
+                    "hidden_states": torch.zeros(
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py:245]: Why u need `self.pp_proxy_tensors_buffer` and why the shape is `[1, hidden_dim]` instead of `['num_tokens, hidden_dim]` here

**Review by Oasis-Git — COMMENTED** [review]
In general the function is good. However before merge we may need to do some code modification.

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py (hunk 5 above)
Comment (baonudesifeizhai) [reply python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py:245]: changed
